ls1intum/Artemis
# `Development`: Improve docker setup for local version control and continuous integration

### Checklist
#### General


#### Server

#### Changes affecting Programming Exercises

### Motivation and Context
There is no working Docker setup for the local VC CI. As the Local CI needs to start build job containers from within the application to run on the host system, this becomes a bit more technical.

A single-command start of the Local VC CI would simplify the setup and allow for E2E testing.

### Description
I added a new Docker Compose file for the a Docker Local VC CI MySQL setup.
I implemented a way to populate the build job container with the necessary repositories and the build script, that works for both setups: Artemis runs on the host or Artemis runs in a Docker container.

### Steps for Testing


Prerequisites:
- Instructor

1. Create Programming exercise
2. Check whether template and solution build have expected results
3. Participate in Exercise and submit
4. Check whether submission has expected results

#### Docker setup
Prerequisites:
- Local VC CI setup on local machine
- change `services.artemis-app.image: ghcr.io/ls1intum/artemis:pr-7226` in `docker/artemis.yml` so the image of this PR is pulled
- if a `ghcr.io/ls1intum/artemis` image is stored on your machine, make sure to delete it
- **On Unix systems:**

  - in the Docker Compose file `docker/artemis-dev-local-vc-local-ci-mysql.yml` replace the value in `services.artemis-app.group_add` with the docker group ID of your machine
  - you can retrieve the docker gid using the command: `getent group docker | cut -d: -f3`
- **On Windows**:
  - in the environment file `docker/artemis/config/dev-local-vc-local-ci.env` add the following line: `ARTEMIS_CONTINUOUSINTEGRATION_DOCKERCONNECTIONURI="tcp://host.docker.internal:2375"`
- start the Docker setup with `docker compose -f docker/artemis-dev-local-vc-local-ci-mysql.yml up`
- Artemis should be available at `localhost:8080` once running

1. Create Programming exercise
2. Check whether template and solution build have expected results
3. Participate in Exercise and submit
4. Check whether submission has expected results


### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test
